### PR TITLE
avm2: Use custom Error instead of std::error::Error

### PR DESCRIPTION
**Pull request by Aaron1011**
This allows us to invoke catch blocks with the original thrown object, instead of a stringified version of the error.
Major changes in this PR:

* We now define `avm2::Error` as `pub enum Error<'gc> { AvmError(Value<'gc>), RustError(Box<dyn std::error::Error>) }`. This replaces our uses of `Box<dyn std::error::Error>` in the `avm2` package. As a result, we need to add a `'gc` lifetime parameter in almost every method.
* Invoking typeless catch block now uses the `Value` from `Error::AvmError` (typed catch blocks are not yet implemented)
* The `From<&'str>` and `From<String>` impls are kept, allowing `Err("Some error".into()` to continue working. Eventually, we should replace all of these with either explicit Error objects, or some internal error type. However, this PR is already extremely large, so this is left for a follow-up PR.
* Several native methods have been converted to produce `Error::AvmError` with the proper error object thrown by flash. Tests have been updated to test that the object matches what Flash Player throws.

### CODE REVIEW

File: core/src/avm2/object/script_object.rs [hunk 1]
```diff
@@ -189,12 +190,34 @@ impl<'gc> ScriptObjectData<'gc> {
...
             .unwrap_or(false)
         {
-            Err(format!(
-                "Cannot get undefined property {:?} on class {:?}",
-                local_name,
-                self.instance_of()
-            )
-            .into())
+            let class_name = if let Some(cls) = self.instance_of() {
+                // Like `to_qualified_name`, but uses a `.` instead of `::` separate
```
Comment (adrian17) [inline core/src/avm2/object/script_object.rs:194]: let's extract this to a function (should this be a QName method?)

Comment (Aaron1011) [reply core/src/avm2/object/script_object.rs:194]: I'm not sure if this is generally useful - doesn't everything else use the other format?

File: core/src/avm2/globals/ArgumentError.as [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package {
+    public class ArgumentError extends Error {
+        public function ArgumentError(message:String = "", code:* = 0) {
+            super(message, code)
+            this.name = prototype.name
+        }
+    }
+}
+
+ArgumentError.prototype.name = "ArgumentError"
```
Comment (EmperorBale) [inline core/src/avm2/globals/ArgumentError.as:10]: I'm pretty sure this is the same as:
```as
public class ArgumentError extends Error {
    prototype.name = "ArgumentError";

    public function ArgumentError(message:String = "", code:* = 0) { ...
```

Comment (EmperorBale) [reply core/src/avm2/globals/ArgumentError.as:10]: These Error classes should also be dynamic

File: core/src/avm2/object/script_object.rs (hunk 1 above)
Comment (Aaron1011) [reply core/src/avm2/object/script_object.rs]: After further thought, I agree. I've extracted this to a method.

File: core/src/avm2/globals/Error.as [hunk 3]
```diff
@@ -1,6 +1,8 @@
 package {
 	[Ruffle(InstanceAllocator)]
 	public dynamic class Error {
+		prototype.name = "Error"
+
 		public var name:String = "Error";
```
Comment (EmperorBale) [inline core/src/avm2/globals/Error.as:6]: Nit: 
```suggestion
		public var name:String;
```
(we set this property in the initializer)

Comment (Aaron1011) [reply core/src/avm2/globals/Error.as:6]: It's only set on the prototype, not as a field.

File: core/src/avm2/object/script_object.rs [hunk 4]
```diff
@@ -189,12 +189,27 @@ impl<'gc> ScriptObjectData<'gc> {
...
-                local_name,
-                self.instance_of()
-            )
-            .into())
+            let class_name = self
+                .instance_of()
+                .map(|cls| {
+                    cls.inner_class_definition()
+                        .read()
+                        .name()
```
Comment (adrian17) [inline core/src/avm2/object/script_object.rs:197]: note for the future (not right now): could probably extract that to a function

**Review by EmperorBale — APPROVED** [review]
This looks good - I think we should merge this so we can start transitioning the rest of the AVM2 codebase to the new error system in future PR's (I think we should do this in multiple PR's, rather than one large PR)

**Review by adrian17 — APPROVED** [review]

**Pull request merged**